### PR TITLE
non directories in $PATH are now ignored

### LINKED ISSUE #2313: prompt: error: on field 'curr_branch' on running xonsh
Hey, I wanted to try out xonsh and installed it using homebrew, but I see this error every time I get into a `xonsh` shell. It looks like something related to getting the current git branch. 

```
~  xonsh
prompt: error: on field 'curr_branch'
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/usr/local/Cellar/xonsh/0.5.8/libexec/lib/python3.6/site-packages/xonsh/prompt/__amalgam__.py", line 671, in _get_field_value
    value = field_value() if callable(field_value) else field_value
  File "/usr/local/Cellar/xonsh/0.5.8/libexec/lib/python3.6/site-packages/xonsh/prompt/__amalgam__.py", line 486, in current_branch
    has_git = bool(cmds.locate_binary('git', ignore_alias=True))
  File "/usr/local/Cellar/xonsh/0.5.8/libexec/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 7790, in locate_binary
    _ = self.all_commands
  File "/usr/local/Cellar/xonsh/0.5.8/libexec/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 7733, in all_commands
    for cmd in executables_in(path):
  File "/usr/local/Cellar/xonsh/0.5.8/libexec/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 5881, in executables_in
    yield from func(path)
  File "/usr/local/Cellar/xonsh/0.5.8/libexec/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 5849, in _executables_in_posix
    yield from _yield_accessible_unix_file_names(path)
  File "/usr/local/Cellar/xonsh/0.5.8/libexec/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 5829, in _yield_accessible_unix_file_names
    for file_ in scandir(path):
NotADirectoryError: [Errno 20] Not a directory: '/usr/local/bin/jamf'
```

Here are my `xonfig` and `.xonshrc`:

```
r@GOHG ~ (ERROR:curr_branch) $ xonfig
+------------------+-----------------+
| xonsh            | 0.5.8           |
| Python           | 3.6.0           |
| PLY              | 3.9             |
| have readline    | True            |
| prompt toolkit   | None            |
| shell type       | readline        |
| pygments         | None            |
| on posix         | True            |
| on linux         | False           |
| on darwin        | True            |
| on windows       | False           |
| on cygwin        | False           |
| is superuser     | False           |
| default encoding | utf-8           |
| xonsh encoding   | utf-8           |
| encoding errors  | surrogateescape |
+------------------+-----------------+
```

```
$ cat ~/.xonshrc
$XONSH_SHOW_TRACEBACK = True
```

Let me know if you need any other information and/or need help fixing this. Thanks!

### PR DESCRIPTION
**Pull request by gsaga**
A typo on line 97 was the reason behind #2313 

Fixes #2313

### CODE REVIEW

**Comment by scopatz** [comment]
Thanks for putting this in @gsaga!

File: xonsh/commands_cache.py [hunk 1]
```diff
@@ -73,7 +73,7 @@ def get_possible_names(name):
     @property
     def all_commands(self):
         paths = builtins.__xonsh_env__.get('PATH', [])
-        pathset = frozenset(x for x in paths if os.path.isdir(x))
+        pathset = tuple(x for x in paths if os.path.isdir(x))
```
Comment (scopatz) [inline xonsh/commands_cache.py]: This should probably be a set, so that it removes duplicated entries

Comment (scopatz) [reply xonsh/commands_cache.py]: Or maybe a tuple is a better option, but you need to create the tuple after the the set

Comment (gsaga) [reply xonsh/commands_cache.py]: But `frozenset`s can't be reversed. Maybe should just remove the duplications old fashioned way.

Comment (gsaga) [reply xonsh/commands_cache.py]: The problem is that frozensets don't preserve order. Duplications could be removed in EnvPath. Are there any environment path variables which might cause trouble if we remove duplications.

File: news/fixCommandsCacheGen.rst [hunk 2]
```diff
@@ -0,0 +1,7 @@
+**Changed:**
```
Comment (gforsyth) [inline news/fixCommandsCacheGen.rst]: For the news entry file, you should leave the unused headers alone (along with their trailing `None`).  items under a header should have a leading `*` followed by a single space before any text

File: xonsh/commands_cache.py [hunk 3]
```diff
@@ -73,7 +73,14 @@ def get_possible_names(name):
     @property
     def all_commands(self):
         paths = builtins.__xonsh_env__.get('PATH', [])
-        pathset = frozenset(x for x in paths if os.path.isdir(x))
+        def removeDups(p):
```
Comment (gforsyth) [inline xonsh/commands_cache.py]: We (and PEP8) prefer snake_case for function names, can this be `remove_dups` please?

File: xonsh/commands_cache.py [hunk 4]
```diff
@@ -73,7 +73,14 @@ def get_possible_names(name):
...
+            ret = list()
+            for e in p:
+                if e not in ret:
+                    ret.append(e)
+            return ret
+        path = removeDups(path)
+        pathset = tuple(x for x in paths if os.path.isdir(x))
```
Comment (gforsyth) [inline xonsh/commands_cache.py]: Not a blocker, but if this is no longer a set I would prefer if the variable name changed to reflect that.

File: xonsh/commands_cache.py (hunk 3 above)
Comment (scopatz) [reply xonsh/commands_cache.py]: Also, can you please move this function to be a method on the class or to part of the `xonsh.tools` module?  `all_commands` is a performance ciritical pathway and so we don't want to redefine this function each time.

File: xonsh/commands_cache.py (hunk 4 above)
Comment (scopatz) [reply xonsh/commands_cache.py]: Yeah, I agree! Thanks for bringing it up.  maybe `valid_path` is a better name

**Comment by scopatz** [comment]
Thanks for continuing to bang on this @gsaga!

**Comment by gsaga** [comment]
@scopatz @gforsyth I've updated the pull request.

File: xonsh/commands_cache.py [hunk 5]
```diff
@@ -70,12 +70,22 @@ def get_possible_names(name):
...
 
+    @staticmethod
+    def remove_dups(p):
+        ret = list()
+        for e in p:
+            if e not in ret:
+                ret.append(e)
+        return ret
+    
+
```
Comment (scopatz) [inline xonsh/commands_cache.py]: PEP8: I think there is only a single blank line between methods

**Comment by scopatz** [comment]
This looks a ton better @gsaga! Thanks a million.  It looks like Circle is complaining about a couple of small style errors:

```
=================================== FAILURES ===================================
__________________ FLAKE8-check(ignoring E122 E402 F841 E741) __________________
/home/circleci/project/xonsh/commands_cache.py:80:1: W293 blank line contains whitespace
/home/circleci/project/xonsh/commands_cache.py:82:5: E303 too many blank lines (2)
```

Once these are fixed, I think this is ready to go in!

**Comment by gsaga** [comment]
@scopatz @gforsyth The `all_commands`  performs the following checks to decide if the cache is outdated:

1. $PATH has changed
2. aliases dict has changed
3. **Any** of the directories in path have been modified

And if any of these are true the whole command cache is regenerated. Shouldn't the function regenerate commands cache for only those sources which have been modified?

It would probably be better if the EnvPath constructor converts the `str`s and `bytes` passed to it to `pathlib.Path`, cos path entries may contain multiple path separators, for ex:
    '/home/user/Music' != '/home/user//Music'
Converting strings and bytes to `Path` would probably solve that problem.

**Comment by scopatz** [comment]
> I'll get on it.​

That'd be amazing!

**Pull request merged**